Implement batcher for RCP

jaucourt opened the pull request:
https://eaflood.atlassian.net/browse/IWTF-4598

Take the HTTP Batcher in connectors-lib and use it for sending recurring payment requests

jaucourt committed:
Implement batcher for RCP

https://eaflood.atlassian.net/browse/IWTF-4598

Take the HTTP Batcher in connectors-lib and use it for sending recurring payment requests

jaucourt committed:
Revert "Revert amendments to recurring payments job"

This reverts commit 343b73d7081c8cfdc9989747d0256cdee6b63fca.

jaucourt committed:
Merge branch develop into feature/iwtf-4021-govukpay-error-responses-rpjob

jaucourt committed:
Lint fixes

jaucourt committed:
HTTP Request Batcher
- Switched to awaiting fetch promises with allSettled, so errors don't stop processing
- Added maxRequestAttempts to configure max request attempts rather than hard-coding this to 2
- Fixed a bug where if there were fewer requests in the first batch than the maximum batch size, the batch wouldn't process properly

jaucourt committed:
Amend queueRecurringPayment to make payment requests to the correct url, it was using the RCP agreement creation url

jaucourt committed:
Add function to queue up requests for payment status, cascading this to the RCP processor, also added batcher env vars

jaucourt committed:
Retrieving payment status and logging this

jaucourt committed:
Began implementation of checking payment statuses, streamlined tests with helper to implement mock batcher

jaucourt committed:
lint fixes

jaucourt committed:
WIP

jaucourt committed:
All unskipped tests passing and only ten tests left to fix!

jaucourt committed:
All tests passing!

jaucourt committed:
Remove commented code, fix linting, tidy up tests, bugfixes

jaucourt committed:
Merge branch 'develop' into feature/iwtf-4598-implement-batcher-for-rcp

jaucourt committed:
Merge branch 'develop' into feature/iwtf-4598-implement-batcher-for-rcp

jaucourt committed:
Refactored to simplify complex function, then refactored tests to remove all Symbol references in favour of string guids as well as fixing up tests that started failing after the refactor

jaucourt committed:
More simplification to reduce complexity of checkPaymentStatuses

jaucourt committed:
Rename duplicatly named variable

jaucourt committed:
Amend env vars to only what we need and update README to explain them

jaucourt committed:
Added tests to check response json methods are only called once, as well as modifying the method implementation to throw an error if called more than once

irisfaraway commented on packages/connectors-lib/src/govuk-pay-api.js:
```diff
@@ -32,6 +32,23 @@ export const createRecurringPaymentAgreement = async preparedPayment => {
 }
 
+export const queueRecurringPayment = (preparedPayment, batcher) => {
+  batcher.addRequest(process.env.GOV_PAY_API_URL, {
```
Should this be `GOV_PAY_RCP_API_URL`? Same question for `queueRecurringPaymentStatusCheck()` below

jaucourt replied:
Yes, it absolutely should. What I'm wondering about is how, when I tested it, a recurring payment was taken...

irisfaraway commented on packages/connectors-lib/src/__tests__/govuk-pay-api.spec.js:
```diff
@@ -140,14 +148,87 @@ describe('govuk-pay-api-connector', () => {
     })
   })
 
+  describe('queueRecurringPayment', () => {
```
Do we need to test any sort of error handling here, for example if invalid values are passed in or if the batcher itself throws an error?

irisfaraway replied:
(Reply to an older conversation) Still wondering about this

irisfaraway commented on packages/connectors-lib/src/__tests__/http-request-batcher.spec.js:
```diff
@@ -175,11 +175,32 @@ describe('HTTP Request Batcher', () => {
+
+    it('stops retrying requests that received a 429 response once a successful response is received', async () => {
+      const batcher = new HTTPRequestBatcher({ batchSize: 1 })
+      batcher.maxRequestAttempts = 5
```
Is it worth testing with a higher `maxRequestAttempts` than the number of attempts we expect the test to make? To me, it's not immediately clear if the requests stop because we have stopped mocking the 429 response, or because we have hit the cap on requests.

jaucourt replied:
I'm not sure I fully understand, but I'll add a test that mocks out 429 responses below the maxRequestAttempts and then adds a 200 response, then checks that retries were only made until a successful response is received.

If the test didn't stop at maxRequestAttempts, this test would fail because the number of calls to `fetch` would exceed maxRequestAttempts, and the test wouldn't pass.

irisfaraway commented on packages/connectors-lib/src/__tests__/http-request-batcher.spec.js:
```diff
@@ -232,18 +249,35 @@ describe('HTTP Request Batcher', () => {
+
+      await batcher.fetch()
+
+      expect(fetch).toHaveBeenCalledTimes(5)
```
Think it'd be preferable to test which requests have been sent rather than just the number. So maybe adding 5 different requests to the batcher and then checking that each one has been called. That way we know it's 5 different ones being called and not the same one being called again due to a non-200 response.

jaucourt replied:
I would say that's done by 'calls fetch for each item in the queue' on line 64, but perhaps there's value in testing every response, so I'll amend the test to do that.

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -159,47 +202,61 @@ const preparePayment = (agreementId, transaction) => {
+  await batcher.fetch()
+
+  return batcher
+}
```
Could we add some line breaks between the declarations for readability? Same goes for line 248/9.

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -159,47 +202,61 @@ const preparePayment = (agreementId, transaction) => {
+const checkPaymentStatuses = async payments => {
+  const batcher = await queuePaymentStatusChecks(payments)
+  for (let index = 0; index < batcher.responses.length; index++) {
+    const response = batcher.responses[index]
```
Could we do this as a `for ... of` loop instead? Bit nicer to read

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -65,61 +75,94 @@ const fetchDueRecurringPayments = async date => {
+      })
+    } else {
+      const description = response.jsonValue.description || ''
+      if (
+        description.match(new RegExp(`Invalid attribute value: ${paymentToRequest.agreementId}. Agreement (does not exist|must be active)`))
+      ) {
+        console.log(
+          '%s is an invalid agreementId. Recurring payment %s will be cancelled',
```
I think it'd be more helpful to print out a separate `%s is an inactive agreementId` message in that scenario (and then do the cancelling anyway)

jaucourt replied:
I was keeping this as it was in the main branch because it wasn't really included in the scope of the ticket, but it's no big deal to change it.

jaucourt replied:
I've thought about this further, and I'd rather add it to 5190 if we need to. Honestly, I think that whether the agreementId doesn't exist or is inactive - and the former will rarely if ever happen - can be treated as one and the same by this code. If we want to investigate further we can check the agreement id manually, but I doubt that we would, we'd likely just leave it as cancelled.

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -65,61 +75,94 @@ const fetchDueRecurringPayments = async date => {
-      console.log('%s is an invalid agreementId. Recurring payment %s will be cancelled', agreementId, transaction.recurringPayment.id)
-      await salesApi.cancelRecurringPayment(transaction.recurringPayment.id)
+const processPaymentResponses = async (paymentsToRequest, batcher) => {
+  for (let x = 0; x < paymentsToRequest.length; x++) {
```
Can we do a `for ... of` loop instead, or do we need to do it this way to get the index for the batcher?

Honestly just matching based on assuming these arrays are in the same order makes me a little nervous. Is there any other common value we can match on?

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -65,61 +75,94 @@ const fetchDueRecurringPayments = async date => {
+      const transaction = await createNewTransaction(referenceNumber, { agreementId, id })
+      return { agreementId, transaction }
+    })
+  )
```
Some of this feels like it might be a good candidate for abstracting out so it's not all sitting within `requestPayments()` as there's now quite a lot going on in here

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -65,61 +75,94 @@ const fetchDueRecurringPayments = async date => {
+  await processPaymentResponses(paymentsToRequest, batcher)
+
+  return Promise.all(
+    batcher.responses.filter(r => r.status && !isClientError(r.status) && !isServerError(r.status)).map(async r => r.jsonValue)
```
Can we abstract this filtering out into a more descriptive name?

irisfaraway replied:
I also couldn't find any explicit tests for this filtering.

Reply (doing it here as it had marked the code as outdated):
It's tested in 'queues a payment status check for each successful payment response', using the second set of test cases where we have a 4xx and 5xx response.

jaucourt replied:
1) Absolutely
2) They're in 'processRecurringPayments requests payment status for all %i successful payments when responses include %s', the second test case

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -47,11 +50,18 @@ const processRecurringPayments = async () => {
+      recurringPaymentId: duePayment.entity.id,
+      transactionId: payment.reference
+    }
+  })
```
`processRecurringPayments()` is getting pretty long, so might be worth breaking this out into a separate `mapRequestedPayments()` or something

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -40,46 +40,79 @@ jest.mock('@defra-fish/connectors-lib', () => ({
+    return Promise.reject(new TypeError('body used already'))
+  }
+}
+// have to go through all tests and check due payments are matching HTTP batcher responses
```
Not sure what this is referring to - is this a todo note that accidentally got left in?

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -301,101 +330,208 @@ describe('recurring-payments-processor', () => {
 
-    describe('when the error is caused by an invalid agreementId', () => {
-      it('logs out the ids', async () => {
+    it('logs an error for every failed payment', async () => {
```
Might be worth testing with multiple errors also as right now we only test that one message is logged, and also don't test which payment it's being logged for

jaucourt replied:
This is a fair point, but given it isn't really within the scope of the ticket, I'll create a new ticket for the work (IWTF-5190)

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -301,101 +330,208 @@ describe('recurring-payments-processor', () => {
         )
       })
 
       it('cancels the recurring payment', async () => {
```
I'd like to see the invalid/inactive reason included in the test name, like it is for the test above - easier to read in the logs

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -301,101 +330,208 @@ describe('recurring-payments-processor', () => {
-      expect(getPaymentStatus).toHaveBeenCalledTimes(6)
-    })
-  })
+        const [, paymentStatusBatcher] = HTTPRequestBatcher.mock.instances
```
This comma deconstruction syntax wasn't familiar to me so I had to look it up! Wonder if `const paymentStatusBatcher = HTTPRequestBatcher.mock.instances[1]` might have been easier to read? I think that's more in line with what we've done elsewhere. Same goes for the next test down.

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -745,64 +924,70 @@ describe('recurring-payments-processor', () => {
     })
 
     it('does not abort when getPaymentStatus rejects for one payment (allSettled at status stage)', async () => {
```
Since `getPaymentStatus` is no longer being used here, this test may need renaming to clarify what is going on

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -40,46 +40,79 @@ jest.mock('@defra-fish/connectors-lib', () => ({
-const getMockDueRecurringPayment = ({ agreementId = 'test-agreement-id', id = 'abc-123', referenceNumber = '123' } = {}) => ({
+const getPaymentStatusSuccess = (additionalFields = {}) => ({ state: { status: PAYMENT_STATUS.Success }, ...additionalFields })
+const getPaymentStatusFailure = () => ({ state: { status: PAYMENT_STATUS.Failure } })
+const getPaymentStatusError = () => ({ state: { status: PAYMENT_STATUS.Error } })
```
I know these are still being used, but I wonder if they should be renamed for clarity since `getPaymentStatus` is no longer being used in this processor

jaucourt replied:
I think they still describe what they are quite well, even though the getPaymentStatus function has been removed. Honestly can't think of a more descriptive name, let me know if you can.

irisfaraway replied:
(Reply to an older conversation) I was trying to think up names that didn't feel like they could be confused with `getPaymentStatus()`, but then it occurred to me that with `getPaymentStatus()` removed from the processor, we actually aren't using it anywhere else. So maybe just rip that function out completely and leave these mocks as they are instead?

irisfaraway reviewed (changes requested):
Nothing major, it's just a massive PR! There are some places where I'd maybe like to see things abstracted a bit further. Honestly I may take a second pass at reviewing the test suite just to confirm all different scenarios are covered, but submitting the review for now so GitHub doesn't lose it.

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -47,11 +50,18 @@ const processRecurringPayments = async () => {
     return
```
This isn't a new issue to this PR but I don't think we have a test specifically checking that this return works. Maybe one that confirms that if there are no results, none of the subsequent calls to GOV.UK Pay are made?

jaucourt replied:
I'll add this to the testing ticket I created

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -265,31 +289,36 @@ describe('recurring-payments-processor', () => {
     })
 
-    it('logs an error for every failure', async () => {
+    it('logs an error for create transaction failures', async () => {
```
The title of this test also specifies "create transaction failures" but it seems like it's testing errors from both `salesApi.preparePermissionDataForRenewal()` and `salesApi.createTransaction()`. I wonder if it's worth having separate tests for these, or at least updating the name of the test to clarify it is testing multiple types of failure. If we split it out then we could also test the content of the error message, rather than just `expect.any(String)`

Also, should we be checking that if either `salesApi.preparePermissionDataForRenewal()` or `salesApi.createTransaction()` fails for one record, it doesn't derail the process for subsequent records? We have some mocks here to ensure the others are still going through, but we aren't explicitly testing that they do.

jaucourt replied:
Updated the name on this - if you want to add to the testing ticket with the other requirements, then go ahead. We do test that preparePermissionDataForRenewal is called as we'd expect, so I don't have undue worries about this interface.

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -65,61 +75,94 @@ const fetchDueRecurringPayments = async date => {
+    })
+  )
+  logErrors(createTransactionResults, 'Error creating transactions:')
+  const paymentsToRequest = createTransactionResults.filter(ctr => ctr.status === 'fulfilled')
```
I'm not sure we are explicitly testing this filtering anywhere.

jaucourt replied:
Great spot, added a test for this

irisfaraway replied:
(Reply to an older conversation) I can see from the commit history that we are testing that ones that throw errors are not created, but still doesn't look like we are testing against the status being 'fulfilled' - am I missing something?

PB: Can't reply, so editing your comment again! This is tested by 'only creates payments for transactions that are successfully created'.

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -301,101 +330,208 @@ describe('recurring-payments-processor', () => {
+        })
+      })
+
+      it('initiates the batcher to retrieve payment statuses', async () => {
```
I wasn't able to find a corresponding test specifically for calling fetch against the payment request batcher (although we are testing that we add multiple items to its queue on line 1389). Think that might be good to have.

jaucourt replied:
Another good spot, added

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -301,101 +330,208 @@ describe('recurring-payments-processor', () => {
+      })
+
+      await execute()
+      expect(console.error).toHaveBeenCalledWith(expect.stringMatching(/Unexpected response from GOV.UK Pay API\..*/))
```
I'd like to see this test the "Error requesting payments:" message that gets passed into `logErrors()` as well, just so we can be clearer how the error is being called

jaucourt replied:
This has changed now, as I realised logErrors wouldn't do anything with batcher results that failed, because they report as successful. The errors are actually handled in processPaymentResponse

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -65,61 +75,94 @@ const fetchDueRecurringPayments = async date => {
+  for (let x = 0; x < paymentsToRequest.length; x++) {
+    const { value: paymentToRequest } = paymentsToRequest[x]
+    const response = batcher.responses[x]
+    response.jsonValue = response.json ? await response.json() : {}
```
Can we add a test for the empty hash fallback here?

jaucourt replied:
Added ('lack of a json function on the error response is handled gracefully')

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -65,61 +75,94 @@ const fetchDueRecurringPayments = async date => {
+    response.jsonValue = response.json ? await response.json() : {}
+
+    if (isSuccessfulResponse(response.status)) {
+      await salesApi.createPaymentJournal(paymentToRequest.transaction.id, {
```
I couldn't find any tests for what happens if `createPaymentJournal` throws an error

jaucourt replied:
Added to 5190

irisfaraway commented on packages/recurring-payments-job/src/recurring-payments-processor.js:
```diff
@@ -65,61 +75,94 @@ const fetchDueRecurringPayments = async date => {
+          paymentToRequest.agreementId,
+          paymentToRequest.transaction.recurringPayment.id
+        )
+        await salesApi.cancelRecurringPayment(paymentToRequest.transaction.recurringPayment.id)
```
Do we need some testing for what happens if `salesApi.cancelRecurringPayment` raises an error?

jaucourt replied:
Added to 5190

irisfaraway reviewed (changes requested):
Second pass just to look for possible gaps in testing.

jaucourt committed:
Refactor batcher to link requests and responses

jaucourt committed:
Completed batcher refactor like a Renegade Master

jaucourt committed:
Slight refactor to get last element of responses array in a less verbose manner

jaucourt committed:
Amend recurring payments processor to handle changes mad
e to HTTPRequestBatcher

jaucourt committed:
Add reference to request details in batcher

jaucourt committed:
- Amend queueRecurringPayment in govuk-pay-api to add agreementId as a reference to the batcher
- Amend recurringPaymentsProcessor to match up payments with request details by agreement id rather than ordinal position

jaucourt committed:
refactor of RP Processor to simplify more complex functions

jaucourt committed:
Add test for transaction creation failures not being queued for payment

jaucourt committed:
Added explicit test for batcher.fetch being called

jaucourt committed:
Test refactoring and a couple of new tests to round out coverage

jaucourt committed:
Add payment id as the reference for payment status requests

jaucourt committed:
Refactor pairing of payments to response details to be by payment id rather than ordinal position

jaucourt committed:
Merge branch 'develop' into feature/iwtf-4598-implement-batcher-for-rcp

jaucourt committed:
Remove redundant await statement

jaucourt committed:
Merge branch 'feature/iwtf-4598-implement-batcher-for-rcp' of github.com:DEFRA/rod-licensing into feature/iwtf-4598-implement-batcher-for-rcp

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -1321,6 +1438,130 @@ describe('recurring-payments-processor', () => {
+    // console.log(queueRecurringPaymentStatusCheck.mock.calls)
+
+    // expect(queueRecurringPaymentStatusCheck).toHaveBeenNthCalledWith(1, paymentIds[0], expect.any(HTTPRequestBatcher))
+    // expect(queueRecurringPaymentStatusCheck).toHaveBeenNthCalledWith(2, paymentIds[1], expect.any(HTTPRequestBatcher))
```
Some commented-out bits here

irisfaraway commented on packages/recurring-payments-job/README.md:
```diff
@@ -10,7 +10,7 @@ When the RP job runs, all RP entries with an nextDueDate of the current date, a
 | name                           | description                                                     | required | default             | valid                         | notes |
 | ------------------------------ | --------------------------------------------------------------- | :------: | ------------------- | ----------------------------- | ----- |
-| NODE_ENV                       | Node environment                                                |    no    |                     | development, test, production |       |
+| NODE_ENV                       | Node environment                                                |    no    |                     | development, fsh-{env}.       |       |
```
There's an extra full stop at the end here

irisfaraway commented on packages/connectors-lib/src/__tests__/http-request-batcher.spec.js:
```diff
@@ -190,6 +232,21 @@ describe('HTTP Request Batcher', () => {
       expect(fetch).toHaveBeenCalledTimes(maxRequestAttempts)
     })
 
+    // it.each([10, 7])('only retries until a successful response is received when maxRequestAttempts is set to %i', async maxRequestAttempts => {
```
Commented-out code here. I'd probably prefer the commented-out version so we can test the variety, but if we're sticking with just testing 7, could probably simplify the test description so we don't iterate over an array with only one entry

jaucourt replied:
Nope, my mistake, I've put both versions back in

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -366,11 +407,12 @@ describe('recurring-payments-processor', () => {
 
       await execute()
+
       expect(console.error).toHaveBeenCalledWith(expect.stringMatching(/Unexpected response from GOV.UK Pay API\..*/))
```
Is it worth testing that the full error data is logged correctly (with the error message, stringified payload, etc?)

jaucourt replied:
I'm not a fan of testing the full error message, at least in this case, because I think it makes for a brittle test

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -343,20 +382,22 @@ describe('recurring-payments-processor', () => {
         getMockDueRecurringPayment({ referenceNumber: 'fum', agreementId: 'a4' })
       ])
+      const errorResponse = new Error('Payment request failed')
+      errorResponse.reference = 'a2'
```
I'm not sure where we do anything with this?

jaucourt replied:
We do not - I'll delete

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -574,6 +630,15 @@ describe('recurring-payments-processor', () => {
+
+    await execute()
+
+    expect(queueRecurringPayment).toHaveBeenCalledTimes(1)
```
Would prefer to test what we're calling it with, rather than just the number of calls

jaucourt replied:
We do that elsewhere, though - this is literally just to test that queueRecurringPayment isn't called for a transaction failure

irisfaraway replied:
Ahh gotcha. LGTM then!

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -1321,6 +1438,130 @@ describe('recurring-payments-processor', () => {
+
+    await execute()
+
+    expect(salesApi.processRPResult).toHaveBeenCalledWith('test-transaction-id', 'test-payment-id', expect.any(String))
```
I'm not sure this succeeds in testing the 429 for the first batcher, since we are mocking the second batcher to always return a success response with those payment identifiers. At the moment I think we could swap out `paymentResponse` with totally different values and this would still pass.

Maybe if we throw the 429 into `statusResponses` instead, or test that the second batcher is calling the values provided? Although that may already be covered elsewhere. I don't know, it's an awkward one.

jaucourt replied:
I've refactored this slightly so we're using proper guids for the various identifiers, but mocking out the first batcher is what we want - we mimic a 429 response, which the real batcher would retry, so I model that with a subsequent 200 response. The expected behaviour is that the recurring payments processor takes the last (successful) response from the array of responses and then continues with making a payment status response and checks the record is processed. I suppose we could change to simply making sure that the payment status request is done and ignoring the call to processRPResult, but the effect would be the same.

irisfaraway replied:
Yeah that's fair, I can't really think of a better way to handle it!

irisfaraway commented on packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:
```diff
@@ -1321,6 +1438,130 @@ describe('recurring-payments-processor', () => {
+      { status: 200, reference: agreementIds[0], json: getJSONImplementation({ payment_id: 'payment-1', agreement_id: agreementIds[0] }) }
+    ]
+    mockBatchers({
+      // payment responses are intentionally out of order to test matching by agreementId
```
Can we move this comment up to where `paymentResponses` are defined? Would be slightly clearer

jaucourt replied:
Done

irisfaraway reviewed (changes requested):
Refactor looks good. I'm definitely happier with how the matching works now :) No major feedback, just a few suggestions around testing and some commented-out lines. I did also reply to some of my older comments so hopefully that comes through okay (I think we can rip out `getPaymentStatus()` entirely!)

jaucourt committed:
Review fixes

jaucourt committed:
lint

jaucourt committed:
Removed redundant functions

irisfaraway reviewed (approved):
🎺 🎺 🎺 :shipit: 🎉

jaucourt committed:
Update README with correct env vars, fix payment request to use POST request limit not GET request limit, fix log of error so it uses the response detail - there was a bug in the tests

jaucourt committed:
Merge branch 'develop' into feature/iwtf-4598-implement-batcher-for-rcp

jaucourt committed:
chore(release): publish v1.73.0-rc.7-test-iwtf-4598

jaucourt committed:
Amend url for requesting payments and retrieving status back to what it was

jaucourt committed:
chore(release): publish v1.73.0-rc.7-test-iwtf-4598-attempt-2

jaucourt committed:
merge

jaucourt committed:
Updated package-lock files
